Add NxTransferList ordering capability - RSC-649

entroform opened the pull request:
https://issues.sonatype.org/browse/RSC-649

entroform commented:
# Summary of changes:

Added a new `allowReordering` property to NxTransferList component.
If `allowReordering` is true:
- `selectedItems` must receive an Array instead of a Set; otherwise, it will throw an error.
-  You must also pass in both type parameters for `id` and `selectedItems` to NxTransferList, example:
`<NxTransferList<number, number[]> ...`
- When you select an item, the last item you selected appears at the bottom of the second column, instead of following the order of `allItems`.

If `allowReordering` is false:
- It behaves as before and you can must pass in a Set.
- `id` type is default as a string, if it's not, you must pass the type parameter, `<NxTransferList<number> ...`. This is probably a breaking change? I'm going to investigate a way around this.

rpokorny commented on gallery/src/components/NxTransferList/NxTransferListPage.tsx:
```diff
@@ -59,13 +61,31 @@ const NxTransferListPage = () =>
-                when updating this prop, rather than mutating the same <NxCode>Set</NxCode>.
+                A Set, or an Array if <NxCode>allowReordering</NxCode> is set to true, containing the ids of the
+                items which are selected – that is, the ones which should
+                appear on the right side of the component. Be sure to always provide a fresh
+                <NxCode>Set</NxCode> or <NxCode>Array</NxCode> instance
```
Fixes whitespace around the `<NxCode>`
```suggestion
                appear on the right side of the component. Be sure to always provide a
                fresh <NxCode>Set</NxCode> or <NxCode>Array</NxCode> instance
```

rpokorny commented on gallery/src/components/NxTransferList/NxTransferListPage.tsx:
```diff
@@ -59,13 +61,31 @@ const NxTransferListPage = () =>
+                the user selects the item. The last selected item appears at the bottom of the selected list.
+                Furthermore, if this is set to true, <NxCode>selectedItems</NxCode> must be
+                an <NxCode>Array</NxCode> instead of a <NxCode>Set</NxCode> and you must pass in
+                both the<NxCode>id</NxCode> and <NxCode>selectedItems</NxCode> type
```
```suggestion
                both the <NxCode>id</NxCode> and <NxCode>selectedItems</NxCode> type
```

entroform replied:
> 

Removed this line because it's no longer necessary to define the type parameters it seems:
Updated in this commit: d098925532af2c05b55be1e7b1bd141f26e0352a

rpokorny commented on gallery/src/components/NxTransferList/NxTransferListPage.tsx:
```diff
@@ -59,13 +61,31 @@ const NxTransferListPage = () =>
+              <NxTable.Cell>boolean</NxTable.Cell>
+              <NxTable.Cell>No</NxTable.Cell>
+              <NxTable.Cell>false</NxTable.Cell>
+              <NxTable.Cell>
```
This description makes no mention of the buttons allowing the _user_ to reorder the items

entroform replied:
> This description makes no mention of the buttons allowing the _user_ to reorder the items

Yup, updated the description here: d098925532af2c05b55be1e7b1bd141f26e0352a

rpokorny commented on lib/src/components/NxSearchTransferList/NxSearchTransferList.tsx:
```diff
@@ -40,8 +40,8 @@ export default function NxSearchTransferList<T extends string | number>(props: P
-  function onRemoveAll(idsToRemove: Set<T>) {
-    const newAddedItems = reject(({ id }) => idsToRemove.has(id), addedItems);
+  function onRemoveAll(idsToRemove: T[]) {
+    const newAddedItems = reject(({ id }) => idsToRemove.includes(id), addedItems);
```
I wonder if it's worth it to collect the ids into a Set in a local variable first so that each lookup is O(1) not O(n)

entroform replied:
> I wonder if it's worth it to collect the ids into a Set in a local variable first so that each lookup is O(1) not O(n)

Thanks, I think it's worth it: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -19,8 +20,10 @@ const defaultItemsCountFormatter = (kind: string) => (n: number) => `${n} item${
     defaultSelectedItemsCountFormatter = defaultItemsCountFormatter('transferred');
 
-export default function NxTransferList<T extends string | number>(props: Props<T>) {
+export default function NxTransferList
+<T extends string | number = string, P extends Set<T> | T[] = Set<T>>(props: Props<T, P>) {
```
I think this is the first time we've had type parameters this long but the way I'd format this would probably be
```suggestion
export default function NxTransferList<
  T extends string | number = string,
  P extends Set<T> | T[] = Set<T>
>(props: Props<T, P>) {
```

Thoughts?

entroform replied:
> I think this is the first time we've had type parameters this long but the way I'd format this would probably be
> 
> Thoughts?

Thanks! Yeah I like it formatted this way as well!

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
+                .map(item => allItems.find(_item => _item.id === item))
+                .filter(item => typeof item !== 'undefined') as DataItem<T>[]
+          },
+      [allItems, selectedItems]
```
```suggestion
      [allItems, selectedItems, allowReordering]
```

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
-  const groupedItems = useMemo(() => groupBy(item => selectedItems.has(item.id) ? 'selected' : 'available', allItems),
-          [allItems, selectedItems]),
+  const groupedItems = useMemo(() =>
+        !allowReordering
+          ? groupBy(item => selectedItemsArray.includes(item.id) ? 'selected' : 'available', allItems)
+          : {
+            available: allItems.filter(item => !selectedItemsArray.includes(item.id)),
+            selected: selectedItemsArray
+                .map(item => allItems.find(_item => _item.id === item))
+                .filter(item => typeof item !== 'undefined') as DataItem<T>[]
+          },
```
This ternary is kind of backwards
```suggestion
        allowReordering ? {
            available: allItems.filter(item => !selectedItemsArray.includes(item.id)),
            selected: selectedItemsArray
                .map(item => allItems.find(_item => _item.id === item))
                .filter(item => typeof item !== 'undefined') as DataItem<T>[]
         } : groupBy(item => selectedItemsArray.includes(item.id) ? 'selected' : 'available', allItems),
```

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
+          : {
+            available: allItems.filter(item => !selectedItemsArray.includes(item.id)),
+            selected: selectedItemsArray
+                .map(item => allItems.find(_item => _item.id === item))
```
Repeatedly iterating `allItems` is probably slow.  I'd first build a lookup map from id to item and then use that here

entroform replied:
> Repeatedly iterating `allItems` is probably slow. I'd first build a lookup map from id to item and then use that here

Thanks, this is updated here: d098925532af2c05b55be1e7b1bd141f26e0352a

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
+          : {
+            available: allItems.filter(item => !selectedItemsArray.includes(item.id)),
+            selected: selectedItemsArray
+                .map(item => allItems.find(_item => _item.id === item))
```
If you use ramda's `chain` you can combine the mapping and filtering into one operation, and probably avoid the cast too

entroform replied:
> If you use ramda's `chain` you can combine the mapping and filtering into one operation, and probably avoid the cast too

Thanks, this is updated here: d098925532af2c05b55be1e7b1bd141f26e0352a

rpokorny commented:
See https://github.com/sonatype/sonatype-react-shared-components/pull/498 for a way to get rid of the second type parameter

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
+  function onChange(checked: boolean, id: T) {
+    const newSelectedItemsArray = checked
+      ? [...selectedItemsArray, id]
+      : selectedItemsArray.filter(item => item !== id);
```
https://ramdajs.com/docs/#without would be helpful here and probably in a number of other places

entroform replied:
> https://ramdajs.com/docs/#without would be helpful here and probably in a number of other places

Thanks! Updated here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -19,8 +20,10 @@ const defaultItemsCountFormatter = (kind: string) => (n: number) => `${n} item${
     defaultSelectedItemsCountFormatter = defaultItemsCountFormatter('transferred');
 
-export default function NxTransferList<T extends string | number>(props: Props<T>) {
+export default function NxTransferList
```
There's a lot of uses of native array methods like `.map()` and `.filter` here.  I try to be consistent about using ramda's helpers instead of those native methods, _especially_ in files that also use a bunch of other ramda functions.

entroform replied:
> There's a lot of uses of native array methods like `.map()` and `.filter` here. I try to be consistent about using ramda's helpers instead of those native methods, _especially_ in files that also use a bunch of other ramda functions.

Refactored here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +41,63 @@ export default function NxTransferList<T extends string | number>(props: Props<T
-  function onUnselectAll(idsToRemove: Set<T>) {
-    const newSelectedItems = new Set<T>(selectedItems);
+  function onReorderItem(id: T, direction: 1 | -1) {
+    const itemIndex = selectedItemsArray.findIndex(item => item === id);
```
IMO a number of things would be cleaner if `TransferListItem` were aware of its index and communicated that index to `onReorderItem` rather than us having to look it up here

entroform replied:
> IMO a number of things would be cleaner if `TransferListItem` were aware of its index and communicated that index to `onReorderItem` rather than us having to look it up here

Yeah I agree, I've updated it here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -93,10 +122,16 @@ export default function NxTransferList<T extends string | number>(props: Props<T
 
 NxTransferList.propTypes = propTypes;
+
+NxTransferList.defaultValues = {
```
When I try to google this ('react "defaultValues"') all I get are results for some third-party form hook.  Are you sure this is a thing? Even if so, it's not something that we use anywhere else.

entroform replied:
> When I try to google this ('react "defaultValues"') all I get are results for some third-party form hook. Are you sure this is a thing? Even if so, it's not something that we use anywhere else.

Removed it here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.tsx:
```diff
@@ -27,11 +39,36 @@ function TransferListItem<T extends string | number = string>(props: TransferLis
-        <input className="nx-transfer-list__checkbox" type="checkbox" checked={checked} onChange={onChange} />
-        <span>{displayName}</span>
-      </label>
+      <div className={classnames('nx-transfer-list__item',
+          { 'nx-transfer-list__item--with-reordering': !!showReorderingButtons })}>
```
Since this `classnames` call is decently complex and multiple lines long, I'd prefer it be pulled out to a variable

entroform replied:
> Since this `classnames` call is decently complex and multiple lines long, I'd prefer it be pulled out to a variable

Moved it to its own variable here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny replied:
> Moved it to its own variable here: 774e970

I meant move the `classnames` invocation itself.  Like `const classes = classnames('foo', className, { ... });`

That's how we typically do it

entroform replied:
> > Moved it to its own variable here: [774e970](https://github.com/sonatype/sonatype-react-shared-components/commit/774e9702b6a1123bac0705354974edd6087775a5)
> 
> I meant move the `classnames` invocation itself. Like `const classes = classnames('foo', className, { ... });`
> 
> That's how we typically do it

Updated it here: a233e5ac88b0924c468d44ba1c0abc2a8d5a7083 👍

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -118,6 +127,43 @@ $-visible-item-count: 10;
 
+.nx-transfer-list__item--with-reordering {
+  .nx-transfer-list__select {
+    border-radius: var(--nx-border-radius);
```
Isn't this redundant?

entroform replied:
> Isn't this redundant?

Yup! Removed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.tsx:
```diff
@@ -27,11 +39,36 @@ function TransferListItem<T extends string | number = string>(props: TransferLis
+        { showReorderingButtons && (
+          <NxButtonBar className="nx-transfer-list__button-bar">
+            <NxButton variant="icon-only"
+                      className={classnames('nx-transfer-list__button',
```
Also pull out these `classnames` calls into local vars

entroform replied:
> Also pull out these `classnames` calls into local vars

Moved it to its own variable here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.tsx:
```diff
@@ -27,11 +39,36 @@ function TransferListItem<T extends string | number = string>(props: TransferLis
+            <NxButton variant="icon-only"
+                      className={classnames('nx-transfer-list__button',
+                          'nx-transfer-list__button--move-up', {
+                            'disabled': isTopItem
```
I think we probably want to disable these via attr, not classname.  There's no reason for them to be focusable when disabled I don't think - the tooltips on these aren't helpful on the disabled ones.

entroform replied:
> I think we probably want to disable these via attr, not classname. There's no reason for them to be focusable when disabled I don't think - the tooltips on these aren't helpful on the disabled ones.

That's a good point: I updated it here: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -85,23 +85,32 @@ $-visible-item-count: 10;
-  border: var(--nx-border-default);
   border-radius: var(--nx-border-radius);
+  display: flex;
+  justify-content: space-between;
```
What's this for?

entroform replied:
> What's this for?

Forgot to remove that: fixed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -85,23 +85,32 @@ $-visible-item-count: 10;
+  display: flex;
+  justify-content: space-between;
+  margin: var(--nx-spacing-2x) 0;
+  padding: 0;
```
Seems redundant

entroform replied:
> Seems redundant

Yup! Removed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -85,23 +85,32 @@ $-visible-item-count: 10;
 }
 
 .nx-transfer-list__item {
   @include nx-container-helpers.container-horizontal;
```
Probably no longer necessary

entroform replied:
> Probably no longer necessary

Removed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -85,23 +85,32 @@ $-visible-item-count: 10;
+  border-radius: var(--nx-border-radius);
+  border: var(--nx-border-default);
   cursor: pointer;
   display: block;
```
Not needed

entroform replied:
> Not needed

Removed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -85,23 +85,32 @@ $-visible-item-count: 10;
   display: block;
-  margin: var(--nx-spacing-2x) 0;
   padding: var(--nx-spacing-2x);
+  width: 100%;
```
This isn't really accurate, I'd rather use `flex: auto`

entroform replied:
> This isn't really accurate, I'd rather use `flex: auto`

Thanks! I updated it here: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -118,6 +127,43 @@ $-visible-item-count: 10;
+$-button-size: 32px;
+
+.nx-transfer-list__button {
+  height: $-button-size;
+  padding: 0;
+  margin: 0;
+  width: $-button-size;
```
These are supposed to be our usual small-style icon-only-buttons right? The right way to make that is with the `small-icon-btn` mixin in `_nx-button-helpers.scss`

entroform replied:
> These are supposed to be our usual small-style icon-only-buttons right? The right way to make that is with the `small-icon-btn` mixin in `_nx-button-helpers.scss`

I gave that mixin a try but the dimension is different from the spec and adds an additional few pixels to the item height. 🤔 

<img width="911" alt="Screen Shot 2021-09-24 at 12 29 30 PM" src="https://user-images.githubusercontent.com/10160030/134709683-ae660db5-dbc9-4af1-ac2a-ef397e22b3d9.png">

<img width="243" alt="Screen Shot 2021-09-24 at 12 32 38 PM" src="https://user-images.githubusercontent.com/10160030/134709846-6497789e-e718-4c3d-b59f-af9e18e8a76c.png">

rpokorny replied:
> I gave that mixin a try but the dimension is different from the spec and adds an additional few pixels to the item height.

I doubt it was deliberate that these buttons be smaller than all of our other small icon-only buttons.  You should check with @alexalyons126 about that.  For reference, generally our small icon-only buttons are 34x34 including the border.  Looks like this mockup has them at 32x32.

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -118,6 +127,43 @@ $-visible-item-count: 10;
+  &.disabled {
+    &,
+    &:hover,
+    &:active {
```
Focus still shows up

entroform replied:
> Focus still shows up

Fixed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/__tests__/TransferListHalf.test.tsx:
```diff
@@ -80,7 +80,7 @@ describe('TransferListHalf', function() {
   });
 
-  it('passes onMoveAll a set containing every id if no filterValue is set', function() {
+  it('passes onMoveAll a set containing every id if no filterValue is an array', function() {
```
This change doesn't make sense.  "set" here was the verb - to _set_ a value.

entroform replied:
> This change doesn't make sense. "set" here was the verb - to _set_ a value.

ops! Fixed it here: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -118,6 +127,43 @@ $-visible-item-count: 10;
+
+$-button-size: 32px;
+
+.nx-transfer-list__button {
```
IMO this class doesn't need to exist, you can just select `.nx-btn` within `.nx-transfer-list__button-bar`

entroform replied:
> IMO this class doesn't need to exist, you can just select `.nx-btn` within `.nx-transfer-list__button-bar`

Sounds good! I removed it here: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

rpokorny commented on lib/src/components/NxTransferList/__tests__/TransferListHalf.test.tsx:
```diff
@@ -273,7 +277,87 @@ describe('TransferListHalf', function() {
+      expect(thirdItemMoveDownButton).toExist();
+      expect(thirdItemMoveUpButton.hasClass('disabled')).toBeFalsy();
+      expect(thirdItemMoveDownButton.hasClass('disabled')).toBeTruthy();
     });
```
Test that they have the right icons too

entroform replied:
> Test that they have the right icons too

Added tests for those in this commit: 774e9702b6a1123bac0705354974edd6087775a5

rpokorny replied:
> Added tests for those in this commit: 774e970

That test's name seems like it's copied from somewhere else

entroform replied:
> > Added tests for those in this commit: [774e970](https://github.com/sonatype/sonatype-react-shared-components/commit/774e9702b6a1123bac0705354974edd6087775a5)
> 
> That test's name seems like it's copied from somewhere else

Yup! That was fixed here: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

entroform commented on gallery/src/components/NxStatefulTransferList/NxStatefulTransferListExample.tsx:
```diff
@@ -13,5 +13,5 @@ const items = map(i => ({ id: i, displayName: `Item ${i}` }), range(1, 101));
   const [selectedItems, setSelectedItems] = useState<Set<number>>(new Set());
 
-  return <NxStatefulTransferList allItems={items} selectedItems={selectedItems} onChange={setSelectedItems} />;
+  return <NxStatefulTransferList<number> allItems={items} selectedItems={selectedItems} onChange={setSelectedItems} />;
```
```suggestion
  return <NxStatefulTransferList allItems={items} selectedItems={selectedItems} onChange={setSelectedItems} />;
```

entroform commented on lib/src/components/NxTransferList/__tests__/TransferListHalf.test.tsx:
```diff
@@ -273,7 +277,112 @@ describe('TransferListHalf', function() {
+    expect(onReorderItem).toHaveBeenCalledWith(1, -1);
+  });
+
+  it('sets the .nx-transfer-list__move-all icon to an x in a circle if isSelected, otherwise a plus in a circle',
```
Ops, need to update this!

entroform replied:
> Ops, need to update this!

Fixed: 8da7e29ba73263e6ce9f96e12d36ba9915305d0f

entroform commented:
> See #498 for a way to get rid of the second type parameter

Thanks! I updated the PR with your suggestion: cad35f0a67ae32b750ecf7df8b90a31bdb2e7143

rpokorny commented on gallery/src/components/NxTransferList/NxTransferListPage.tsx:
```diff
@@ -160,7 +179,8 @@ const NxTransferListPage = () =>
               <NxTable.Cell>
                 Callback for when the user selects or unselects one or more items. Receives the
-                new <NxCode>Set</NxCode> of selected ids as a parameter.
+                new <NxCode>Set</NxCode>, or <NxCode>Array</NxCode> if <NxCode>AllowReordering</NxCode> is true,
```
```suggestion
                new <NxCode>Set</NxCode>, or <NxCode>Array</NxCode> if <NxCode>allowReordering</NxCode> is true,
```

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +40,68 @@ export default function NxTransferList<T extends string | number>(props: Props<T
 
-  const groupedItems = useMemo(() => groupBy(item => selectedItems.has(item.id) ? 'selected' : 'available', allItems),
-          [allItems, selectedItems]),
+  const allItemsIdToItemlookUp = groupBy(item => item.id.toString(), allItems);
```
```suggestion
  const allItemsIdToItemLookUp = groupBy(item => item.id.toString(), allItems);
```

rpokorny commented on lib/src/components/NxTransferList/NxTransferList.tsx:
```diff
@@ -38,39 +40,68 @@ export default function NxTransferList<T extends string | number>(props: Props<T
+        allowReordering
+          ? {
+            available: reject(item => selectedItemsSet.has(item.id), allItems),
+            selected: chain<T, DataItem<T>>(item => allItemsIdToItemlookUp[item.toString()], selectedItemsArray)
```
For me it's working without the type annotation

```suggestion
            selected: chain(item => allItemsIdToItemlookUp[item.toString()], selectedItemsArray)
```

rpokorny commented:
Re-ordering doesn't seem to work when a filter is active.  The buttons just do nothing.

rpokorny commented on lib/src/components/NxTransferList/__tests__/NxTransferList.test.tsx:
```diff
@@ -22,7 +22,9 @@ describe('NxTransferList', function() {
       },
-      getShallow = getShallowComponent<Props<string | number>>(NxTransferList, minimalProps);
+      // eslint-disable-next-line no-spaced-func
+      getShallow = getShallowComponent<Props<string | number, Set<number | string> | (string | number)[]>>
```
If I'm not mistaken this change is no longer valid

entroform replied:
> If I'm not mistaken this change is no longer valid

Yup! This was fixed here: c1f9f1f29542985e93f7345f1af4eaf152f3e952

rpokorny commented on lib/src/components/NxTransferList/__tests__/NxTransferList.test.tsx:
```diff
@@ -269,4 +271,101 @@ describe('NxTransferList', function() {
+      expect(secondHalf).toHaveProp('items', items);
+    });
+
+    it('fires onChange with a new Array with the clicked item added at the end'
```
```suggestion
    it('fires onChange with a new Array with the clicked item added at the end '
```

rpokorny commented on lib/src/components/NxTransferList/__tests__/NxTransferList.test.tsx:
```diff
@@ -269,4 +271,101 @@ describe('NxTransferList', function() {
   });
+
+  it('throw an error when passing a Set to selectedItems when allowReordering is true', function() {
+    expect(() => getShallow({ allowReordering: true }))
```
This should no longer compile.  If it does, the props changes didn't work as expected

entroform replied:
> This should no longer compile. If it does, the props changes didn't work as expected

Sadly this did compile 😭

rpokorny replied:
> Sadly this did compile sob

Actually that might just be a problem with `getShallow`.  If I change `NxTransferListWithReorderingExample` to use a Set instead of an Array, it fails appropriately

rpokorny commented on lib/src/components/NxTransferList/types.ts:
```diff
@@ -14,53 +14,69 @@ export type FilterFn<T extends string | number = string> = (d: DataItem<T>[]) =>
-    displayName: PropTypes.string.isRequired
-  }).isRequired).isRequired,
-  selectedItems: PropTypes.instanceOf<Set<string | number>>(Set).isRequired,
+// allItems, selectedItems, onAvailableItemsFilterChange, onSelectedItemsFilterChange
+// are excluded in the propTypes due to clash with parametric Props;
```
We should check whether this is still the case now that the types have been redone

entroform replied:
> We should check whether this is still the case now that the types have been redone

I was able to bring 2 back without errors in this commit:
ced6b8e23042c0886a1255efc67fb2f75f92812f

entroform commented:
> Re-ordering doesn't seem to work when a filter is active. The buttons just do nothing.

Thanks for finding this!
I've got a working implementation in this commit: c1f9f1f29542985e93f7345f1af4eaf152f3e952
I'm going to take some time to clean it up and investigate the props change issue.

rpokorny commented:
> Thanks for finding this!
I've got a working implementation in this commit: c1f9f1f
Further cleaned up: 5e5713b

I would say that part of fixing this requires having a discussion with design about how it should behave in this case, because it's not at all clear to me what would be most expected.  What it does currently is it swaps the reordered item with its adjacent visible item. 

For an example, observe the following screenshots.  First, I start with a few selected items and no filter:
![image](https://user-images.githubusercontent.com/3630091/135306489-2811b09f-dec4-4ef3-910f-a7196e254a9e.png)

Then, I filter on "2" which hides all but two of the selected items:
![image](https://user-images.githubusercontent.com/3630091/135306545-3df8cb48-c477-4d3f-bb85-54890f012cd7.png)

Then, I reorder the first item downwards:
![image](https://user-images.githubusercontent.com/3630091/135306614-500238cc-ec1c-42ff-aebf-9a811d041e1e.png)

Then, I clear the filter and observe that the item I moved down hasn't truly just moved down one spot, but has moved down three spots _and the other item that had been visible when filtering has moved **up** three spots_ (they swapped places):
![image](https://user-images.githubusercontent.com/3630091/135306787-68a83da9-075e-4a44-8ada-1803743607c8.png)

entroform commented:
> > Thanks for finding this!
> > I've got a working implementation in this commit: [c1f9f1f](https://github.com/sonatype/sonatype-react-shared-components/commit/c1f9f1f29542985e93f7345f1af4eaf152f3e952)
> > Further cleaned up: [5e5713b](https://github.com/sonatype/sonatype-react-shared-components/commit/5e5713b92df85bb18cc4564c34d3f3de627c58a2)
> 
> I would say that part of fixing this requires having a discussion with design about how it should behave in this case, because it's not at all clear to me what would be most expected. What it does currently is it swaps the reordered item with its adjacent visible item.
> 
> For an example, observe the following screenshots. First, I start with a few selected items and no filter: ![image](https://user-images.githubusercontent.com/3630091/135306489-2811b09f-dec4-4ef3-910f-a7196e254a9e.png)
> 
> Then, I filter on "2" which hides all but two of the selected items: ![image](https://user-images.githubusercontent.com/3630091/135306545-3df8cb48-c477-4d3f-bb85-54890f012cd7.png)
> 
> Then, I reorder the first item downwards: ![image](https://user-images.githubusercontent.com/3630091/135306614-500238cc-ec1c-42ff-aebf-9a811d041e1e.png)
> 
> Then, I clear the filter and observe that the item I moved down hasn't truly just moved down one spot, but has moved down three spots _and the other item that had been visible when filtering has moved **up** three spots_ (they swapped places): ![image](https://user-images.githubusercontent.com/3630091/135306787-68a83da9-075e-4a44-8ada-1803743607c8.png)

Yeah, thanks for pointing that out. I assumed (which I shouldn't have) and thought that was the natural expected behaviour. I will bring it up with design in the #react-components channel.

entroform commented:
> > Thanks for finding this!
> > I've got a working implementation in this commit: [c1f9f1f](https://github.com/sonatype/sonatype-react-shared-components/commit/c1f9f1f29542985e93f7345f1af4eaf152f3e952)
> > Further cleaned up: [5e5713b](https://github.com/sonatype/sonatype-react-shared-components/commit/5e5713b92df85bb18cc4564c34d3f3de627c58a2)
> 
> I would say that part of fixing this requires having a discussion with design about how it should behave in this case, because it's not at all clear to me what would be most expected. What it does currently is it swaps the reordered item with its adjacent visible item.
> 
> For an example, observe the following screenshots. First, I start with a few selected items and no filter: ![image](https://user-images.githubusercontent.com/3630091/135306489-2811b09f-dec4-4ef3-910f-a7196e254a9e.png)
> 
> Then, I filter on "2" which hides all but two of the selected items: ![image](https://user-images.githubusercontent.com/3630091/135306545-3df8cb48-c477-4d3f-bb85-54890f012cd7.png)
> 
> Then, I reorder the first item downwards: ![image](https://user-images.githubusercontent.com/3630091/135306614-500238cc-ec1c-42ff-aebf-9a811d041e1e.png)
> 
> Then, I clear the filter and observe that the item I moved down hasn't truly just moved down one spot, but has moved down three spots _and the other item that had been visible when filtering has moved **up** three spots_ (they swapped places): ![image](https://user-images.githubusercontent.com/3630091/135306787-68a83da9-075e-4a44-8ada-1803743607c8.png)

Updated the PR so the move up and down buttons are now disabled when there is filter value:
306a01e484ebaabee1933fae8e81d1b9b158c3a5

rpokorny commented:
> Updated the PR so the move up and down buttons are now disabled when there is filter value:
306a01e

Should the buttons have a tooltip to explain why they're disabled in this case?

entroform commented:
> Should the buttons have a tooltip to explain why they're disabled in this case?

I think that's a good idea: I've updated it in this commit:
2fc88b083dc2fb00adaa0466d58bddb9a2d2b106

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.tsx:
```diff
@@ -48,6 +48,11 @@ function TransferListItem<T extends string | number = string>(props: TransferLis
       }
   );
 
+  const moveUpButtonTitle = isTopItem
+    ? null : isFilteredItem ? 'Reordering is disabled when filtered' : 'Move Up';
```
IMO nested ternaries are only readable when split over multiple lines such that each line indicates a condition and the result when that condition is true, followed at the end of all the lines by the base case.  Like this:

```
  const moveUpButtonTitle = isTopItem ? null : 
      isFilteredItem ? 'Reordering is disabled when filtered' : 
      'Move Up';
```

rpokorny commented on lib/src/components/NxTransferList/__tests__/TransferListHalf.test.tsx:
```diff
@@ -273,7 +277,149 @@ describe('TransferListHalf', function() {
+    moveUpButton.simulate('click');
+
+    expect(onReorderItem).not.toHaveBeenCalled();
+  });
```
Needs a test for the title logic

entroform replied:
> Needs a test for the title logic

Added test here: 5759f98d182e17a0e9d67472f4f8f68b2f8af0e5

rpokorny replied:
> Added test here: 5759f98

:+1: that'll need to be updated though as the approach has changed slightly

entroform replied:
> > Added test here: [5759f98](https://github.com/sonatype/sonatype-react-shared-components/commit/5759f98d182e17a0e9d67472f4f8f68b2f8af0e5)
> 
> 👍 that'll need to be updated though as the approach has changed slightly

Yup test updated: 885f3cb8f599364a04751640783fbcad7193f3df 😄

rpokorny commented:
Something's wrong with the tooltips; they don't disappear after the mouse leaves.
![image](https://user-images.githubusercontent.com/3630091/135857430-c5b8cac2-1514-48f7-bab4-6e5e7b1e0587.png)

entroform commented:
> Something's wrong with the tooltips; they don't disappear after the mouse leaves. ![image](https://user-images.githubusercontent.com/3630091/135857430-c5b8cac2-1514-48f7-bab4-6e5e7b1e0587.png)

This issue didn't show up on Firefox strangely:
Fixed: 5759f98d182e17a0e9d67472f4f8f68b2f8af0e5

I tried wrapping NxButton in span/div + NxTooltip but the issue is still there. The only way I got it to work is using the 'disabled' class. Let me know if you know a better way around it 😭.

rpokorny commented:
> I tried wrapping NxButton in span/div + NxTooltip but the issue is still there. The only way I got it to work is using the 'disabled' class. Let me know if you know a better way around it sob.

How about like this: https://github.com/sonatype/sonatype-react-shared-components/pull/512

entroform commented:
> > I tried wrapping NxButton in span/div + NxTooltip but the issue is still there. The only way I got it to work is using the 'disabled' class. Let me know if you know a better way around it sob.
> 
> How about like this: #512

Ooo thanks, that looks great. Grouping both buttons with tooltip also makes a lot of sense. Thanks for your help!

rpokorny commented on lib/src/components/NxTransferList/TransferListHalf.scss:
```diff
@@ -118,6 +123,45 @@ $-visible-item-count: 10;
+    width: $-button-size;
+
+    &[disabled],
+    &.disabled {
```
I don't think this is needed anymore

entroform replied:
Yup! Updated here along with the test: 885f3cb8f599364a04751640783fbcad7193f3df

rpokorny commented:
Oh whoops, I rescind my approval as I've just noticed that an important thread is still unresolved: https://github.com/sonatype/sonatype-react-shared-components/pull/488#discussion_r714987126

entroform commented:
> Oh whoops, I rescind my approval as I've just noticed that an important thread is still unresolved: [#488 (comment)](https://github.com/sonatype/sonatype-react-shared-components/pull/488#discussion_r714987126)

Thanks! Just confirming that will be updated in this ticket: https://issues.sonatype.org/browse/RSC-735
